### DISCUSSION TITLE
Text behind drawing

### DISCUSSION THREAD

**Discussion by caiocesarrm** [pymupdf/PyMuPDF#1775] [Looking for help] [answered]
Hi there!

I know that text visibility is an hard problem, so i'll go direct to the point.

This is only a example, i'm searching for a general solution if possible. the blue lines in the table are text that i'm selecting to show that this text is behind the drawing. I believe that i can't reddact because it will erase the table text. Rewriting the entire pdf (including drawing, images, etc...) page by page to new pdf, but excluding text rect that is entirely behind a drawing (detect if it's behing by rendering order) is doable? If so, what is needed to reconstruct a page, copy drawings, images and text?

Thanks in advance!

The page:
[page.pdf](https://github.com/pymupdf/PyMuPDF/files/8985256/page.pdf)

hidden (not 3 Tr) text selected
![image](https://user-images.githubusercontent.com/41218670/175783772-1bdd4af6-80b7-49c9-9e9e-0437ff0a5472.png)
text detected in the page:
![image](https://user-images.githubusercontent.com/41218670/175783831-19e9e1f5-f13d-4c80-87ac-ff14b719aa9b.png)
lines 2,3,4,6,7 with the "invisible" (not 3 tr text) text being extracted in the table extractor.
![image](https://user-images.githubusercontent.com/41218670/175783989-7bc0641a-58b0-4cf3-8ff8-17fd57ad5cec.png)

**Comment by JorjMcKie** [comment, marked as answer]
I am not sure I understand what you actually want:

* either extract text whether or not visible?
* or find out whether text is invisible - even if not marked as such by the Tr 3 parameter?

If the first  aspect:
This is a no-brainer: text extraction always works if it actually **_is_** text (which means: not everything looking like text is text)

The second aspect is a little trickier, because text extraction will not tell you about visibility.
But there are ways to still fnd things out:
`page.get_bboxlog()` returns a list of rectangles of stuff being shown on a page, together with the type of content wrapped by the rect: text (including whether Tr 3), images or drawings.
The sequence in the list represents the sequence in which the objects are triggered for display.
So you can take a text bbox, determine which index of the corresponding item in the bboxlog is containing it, then check whether a later image, drawing, or even other text, ... covers the same space.

But note that a full blown reliable analysis is still not guarantied: a covering image may be transparent, a drawing in the same place need not imply it makes things underneath invisible, etc.

**Comment by caiocesarrm** [comment]
Thanks!

i want to remove the "invisible" text from the page, because when i use tools like camelot-py the table extraction fail or return the table with cell values contanimed with this "invisible" text (print 1 is the table, and the last print show that the cell values are polluted with strings that are behind the table draw.

But your second response probably will help me to rebuild the page without this invisible text, thanks!

**Reply by JorjMcKie** [reply]
You may also want to consider this:
There are Pixmap methods that let you check the number of colors occurring in some area of the page.
So take some text piece like a word and check how many colors are there in its text bbox.
As a reference consult [this](https://pymupdf.readthedocs.io/en/latest/pixmap.html#Pixmap.color_topusage) part of the documentation.
